client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -492,6 +490,9 @@ public void flush() throws SegmentSealedException {
             }
             state.waitForInflight();
             Exceptions.checkNotClosed(state.isClosed(), this);
+            if (state.noSuchSegmentEncountered.get()) {
+                throw new NoSuchSegmentException(segmentName + " is not present due to truncation or deletion.");
```

Comment:
I think this should be thrown as a SegmentSealedException for the benifit of EventStreamWriterImpl's handling.

Reply:
In a scenario where the stream is deleted post writer creation a `flush()`  or `close()` might stuck indefinitely if we throw a SegmentSealedException. This PR changes the behavior of the writer as shown below (Let me know if we should change it )

![behaviour](https://user-images.githubusercontent.com/11114165/40400529-28d38f62-5e60-11e8-8604-62e6459229b7.png)

Reply:
In the case where `flushInternal()` is called, either due to a seal or `flush()` being called, and
it goes down this codepath:
https://github.com/pravega/pravega/pull/2612/files#diff-60053e16af1ab55e04df790f148a7bcaR103
because there are successors to the deleted segment, doesn't this change mean that:
https://github.com/pravega/pravega/pull/2612/files#diff-3c18ee5eb24224b2c2fdcce1e56f97eeR347
Should also be catching `NoSuchSegmentException`, because we want the unacked events to be written to the new segment. 

Yes, you are correct doing this nieevly would result in an infinate loop. We have to think about how to best do that.

I am also worried about duplicates in this case. IE: It is technically possible that there was an unacked append outstanding against the server, and the segment was deleted by retention. Then before the ack or lack there of could be returned to the client the connection dropped. In this case the reconnect will fail because the segment does not exist. So the client will have no way of knowing if the append made it or not. I'm not sure how big of a problem it is because the event even if written would have been immediately deleted. But it is still theoretically possible it could have been read. We could and probably should impose the limitation that truncation will not delete the currently active segment. (And maybe impose some minimum time limit before deleting it so as to avoid this complication)

Reply:
Updated the writer behavior; now `flush()` on receiving `NoSuchSegment` response from SegmentStore will retry sending the inflight events by fetching the deleted segment's successors (simiilar to segment sealed scenario). The updated behavior of writer is shown below.

![image](https://user-images.githubusercontent.com/11114165/40466803-dcbfed38-5f44-11e8-9368-53b76a04843f.png)